### PR TITLE
Make guest username, uid, home directory configurable

### LINKED ISSUE #2037: Faulty installation of Ubuntu (Waiting for the essential requirement 1 of 5: "ssh"): `util.py[WARNING]: Failed to create user admin`
### Description

Steps to reproduce:
`limactl create template://ubuntu --name=ubuntu`
or
`limactl create template://ubuntu-lts --name=ubuntu`
`limactl start ubuntu`

Output:
```
INFO[0000] Using the existing instance "ubuntu"
INFO[0000] QEMU binary "/usr/local/bin/qemu-system-x86_64" seems properly signed with the "com.apple.security.hypervisor" entitlement
INFO[0001] [hostagent] hostagent socket created at /Users/admin/.lima/ubuntu/ha.sock
INFO[0001] [hostagent] Starting QEMU (hint: to watch the boot progress, see "/Users/admin/.lima/ubuntu/serial*.log")
INFO[0001] SSH Local Port: 56893
INFO[0001] [hostagent] Waiting for the essential requirement 1 of 5: "ssh"
INFO[0011] [hostagent] Waiting for the essential requirement 1 of 5: "ssh"
...
```

Obviously, there is a problem with creating the `admin` user in __cloud-init__
Snippet from `~/.lima/ubuntu/serial.log`:
```
[   22.661276] cloud-init[586]: 2023-11-28 05:48:29,178 - util.py[WARNING]: Failed to create user admin
...
[   77.605933] cloud-init[2346]: + loginctl terminate-user admin
[   77.616303] cloud-init[2346]: Failed to look up user admin: No such process
...
[   77.864749] cloud-init[2346]: + grep -q # Lima BEGIN /home/admin.linux/.profile
[   77.867782] cloud-init[2346]: grep: /home/admin.linux/.profile: No such file or directory
[   77.872246] cloud-init[2346]: /mnt/lima-cidata/boot/20-rootless-base.sh: 10: cannot create /home/admin.linux/.profile: Directory nonexistent
[   77.878411] cloud-init[2346]: LIMA| WARNING: Failed to execute /mnt/lima-cidata/boot/20-rootless-base.sh
...
[   77.911972] cloud-init[2346]: + id -g admin
[   77.914698] cloud-init[2346]: id: ‘admin’: no such user
[   77.917552] cloud-init[2346]: + gid=
[   77.919394] cloud-init[2346]: LIMA| WARNING: Failed to execute /mnt/lima-cidata/boot/25-guestagent-base.sh
...
```

Alpine Linux works as expected

### LINKED ISSUE #2117: [User name: `admin`] Waiting for the essential requirement 1 of 4: "ssh"
### Description

I can't get seem to connect via ssh to lima with default using either drivers, output for `vz` below. My local username is `admin` in case that's relevant.

```bash
limactl --version    
limactl version 0.19.1

% sw_vers
ProductName:		macOS
ProductVersion:		14.2.1
BuildVersion:		23C71
```
```
limactl start             
INFO[0000] Using the existing instance "default"        
INFO[0000] Starting the instance "default" with VM driver "vz" 
INFO[0000] [hostagent] hostagent socket created at /Users/admin/.lima/default/ha.sock 
INFO[0000] [hostagent] Starting VZ (hint: to watch the boot progress, see "/Users/admin/.lima/default/serial*.log") 
INFO[0000] [hostagent] new connection from  to          
INFO[0001] SSH Local Port: 60022                        
INFO[0001] [hostagent] Waiting for the essential requirement 1 of 2: "ssh" 
INFO[0001] [hostagent] [VZ] - vm state change: running  
INFO[0011] [hostagent] Waiting for the essential requirement 1 of 2: "ssh" 
INFO[0021] [hostagent] Waiting for the essential requirement 1 of 2: "ssh" 
...
```

```bash
{"level":"info","msg":"Waiting for the essential requirement 1 of 2: \"ssh\"","time":"2024-01-08T06:40:33-05:00"}
{"level":"debug","msg":"executing script \"ssh\"","time":"2024-01-08T06:40:33-05:00"}
{"level":"debug","msg":"executing ssh for script \"ssh\": /usr/bin/ssh [ssh -F /dev/null -o IdentityFile=\"/Users/admin/.lima/_config/user\" -o IdentityFile=\"/Users/admin/.ssh/aviculturist\" -o IdentityFile=\"/Users/admin/.ssh/id_bitnym\" -o IdentityFile=\"/Users/admin/.ssh/id_ed25519_moscoviumalchemist\" -o IdentityFile=\"/Users/admin/.ssh/lucr\" -o IdentityFile=\"/Users/admin/.ssh/pickwp_id_rsa\" -o IdentityFile=\"/Users/admin/.ssh/sozot_github_id_rsa\" -o IdentityFile=\"/Users/admin/.ssh/wpspring_github_id_rsa\" -o IdentityFile=\"/Users/admin/.ssh/wpsy_bitbucket_id_rsa\" -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o NoHostAuthenticationForLocalhost=yes -o GSSAPIAuthentication=no -o PreferredAuthentications=publickey -o Compression=no -o BatchMode=yes -o IdentitiesOnly=yes -o Ciphers=\"^aes128-gcm@openssh.com,aes256-gcm@openssh.com\" -o User=admin -o ControlMaster=auto -o ControlPath=\"/Users/admin/.lima/default/ssh.sock\" -o ControlPersist=yes -p 60022 127.0.0.1 -- /bin/bash]","time":"2024-01-08T06:40:33-05:00"}
{"level":"debug","msg":"stdout=\"\", stderr=\"Received disconnect from 127.0.0.1 port 60022:2: Too many authentication failures\\r\\nDisconnected from 127.0.0.1 port 60022\\r\\n\", err=failed to execute script \"ssh\": stdout=\"\", stderr=\"Received disconnect from 127.0.0.1 port 60022:2: Too many authentication failures\\r\\nDisconnected from 127.0.0.1 port 60022\\r\\n\": exit status 255","time":"2024-01-08T06:40:33-05:00"}
```

### PR DESCRIPTION
**Pull request by jandubois**
Also disallows the "admin" username by default (because it is a builtin user in Ubuntu), but can be overridden by setting it explicitly in lima.yaml.

```yaml
user:
  name: john
  comment: John Doe
  uid: 1000
  home: "/home/{{.User}}"
```

Another change is that `{{.User}}` and `{{.UID}}` in host templates now expand to the host values and not the guest values. Note that on Windows the host user is something like `runner\runneradmin` and host uid something like `S-1-5-21-2897396845-3571823323-1290545657-500` (on GitHub runner).

Setting `home` and `comment` does not work yet with `alpine-iso` because `lima-init` hardcodes the home directory to `/home/$USER.linux` and doesn't set the comment at all (implemented in https://github.com/lima-vm/alpine-lima/pull/134).

Fixes #2117
Fixes #2037 
Implements #2622

Replaces #2378

### CODE REVIEW

**Comment by jandubois** [comment]
I recommend using "Hide whitespace" for reviewing this PR, especially for `osutil/user.go`.

**Comment by afbjorklund** [comment]
Should there be an option to avoid "mirroring" the host values, and go with the default ones from the cloud.cfg?

File: templates/default.yaml
Comment (afbjorklund) [inline templates/default.yaml:136]: Should this be using {{.OS}} perhaps? (for Linux, but lowercase the value). Only makes a difference for FreeBSD

**Comment by afbjorklund** [comment]
> * The code also implements support for setting `user.gid` but doesn't document it because cloud-init has no way to set it. We'll have to implement this ourselves in our ci scripts (if we want it).

In terms of features, it probably would be more useful to have "groups" (extra groups) than "gid" (not used so much)

File: templates/default.yaml
Comment (AkihiroSuda) [inline templates/default.yaml:130]: `user.username` stutters

File: templates/default.yaml
Comment (AkihiroSuda) [inline templates/default.yaml:137]: `{{.Home}}` does not have `Dir` suffix

File: templates/default.yaml
Comment (AkihiroSuda) [inline templates/default.yaml:126]: Probably this should be in the `FURTHER ADVANCED CONFIGURATION`  section

Comment (AkihiroSuda) [inline templates/default.yaml:126]: Can we have a test in https://github.com/lima-vm/lima/blob/master/hack/test-templates/test-misc.yaml

File: templates/default.yaml
Comment (afbjorklund) [reply templates/default.yaml:130]: In the code we have `user.Username` because `user.Name` is the GECOS, but the yaml can be simplified

File: pkg/osutil/user.go
Comment (alexandear) [inline pkg/osutil/user.go:95]: `var cache struct {` was good, because no need to export `Cache`. I propose to revert. Maybe we can use `init()` functions to initialize this struct.

File: templates/default.yaml
Comment (afbjorklund) [reply templates/default.yaml:130]: Can add a separate PR for the gecos, since it seems to be used in cloud.cfg (and simple enough in ci)

File: pkg/must/must.go
Comment (afbjorklund) [inline pkg/must/must.go:3]: I feel this package to be a bit obscure, and I don't see why it should not be handled like any other error

File: templates/default.yaml
Comment (afbjorklund) [reply templates/default.yaml:130]: * #2830

File: pkg/osutil/user.go
Comment (jandubois) [reply pkg/osutil/user.go]: I'm just removing the struct altogether; it doesn't really provide anything but some namespacing that we don't really need either.

```go
var (
	once     = new(sync.Once)
	limaUser *user.User
	warnings []string
)
```

File: templates/default.yaml
Comment (jandubois) [reply templates/default.yaml]: done

File: templates/default.yaml
Comment (jandubois) [reply templates/default.yaml]: Ok, will go with `name`, `home`, and `uid` for now, but what should we use for `gecos`?

File: templates/default.yaml
Comment (jandubois) [reply templates/default.yaml]: Yeah, not really part of this PR.

File: templates/default.yaml
Comment (jandubois) [reply templates/default.yaml]: I've changed `homeDir` → `home`.

**Comment by jandubois** [comment]
> Should there be an option to avoid "mirroring" the host values, and go with the default ones from the cloud.cfg?

I don't know how we would set up ssh without knowing the username. If we can figure this out, then I think we should support this.

**Comment by jandubois** [comment]
> In terms of features, it probably would be more useful to have "groups" (extra groups) than "gid" (not used so much)

I have removed all code related to `gid`, so we can figure out the correct schema in due time.

File: pkg/limayaml/limayaml.go
Comment (AkihiroSuda) [inline pkg/limayaml/limayaml.go:89]: The go name and the yaml name are inconsistent.

I also feel that the short name should be just called "Name" and the gecos field should be called "FullName" or maybe "Comment".

"GECOS" is a remnant from 1960s and no longer meaningful word.

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
The go name and the yaml name are inconsistent.

File: pkg/limayaml/limayaml.go
Comment (jandubois) [reply pkg/limayaml/limayaml.go]: I renamed it to `fullName`, but just noticed the `comment` suggestion, and will change to that one instead.

In the CI template we have `Name` for the instance name, and `User` for the user name, so `FullName` does not really fit there (so I left it at `GECOS`), but `Comment` works fine.

File: templates/default.yaml
Comment (AkihiroSuda) [inline templates/default.yaml:286]: Do you want to add `fullName` too?

File: pkg/limayaml/limayaml.go
Comment (AkihiroSuda) [inline pkg/limayaml/limayaml.go:88]: The go name and the yaml name should be synced

**Comment by jandubois** [comment]
I'm wondering if `{{.Comment}}` should be supported in host/guest templates. It would be easy to add, but I can't think of an actual use-case.

**Comment by jandubois** [comment]
I believe I've addressed all feedback now, PTAL

There are 2 open issues that should probably be done "later":

* Optionally don't add an additional user at all (but how do we set up ssh keys?)
* Add `{{.Comment}}` to host and guest templates (but why?)

File: pkg/limayaml/limayaml.go
Comment (AkihiroSuda) [inline pkg/limayaml/limayaml.go:91]: Why not `*int`?

Comment (jandubois) [reply pkg/limayaml/limayaml.go:91]: No particular reason, I can change it.

Comment (jandubois) [reply pkg/limayaml/limayaml.go:91]: @AkihiroSuda Can we leave this for a later refactoring PR? I would also want to combine this type with https://github.com/lima-vm/lima/blob/9fe06aa4cc7ac7d1d6b72787301753cc324ae4d8/pkg/osutil/user.go#L17-L24

But I feel like the current PR is already big enough.

File: pkg/cidata/template_test.go
Comment (AkihiroSuda) [inline pkg/cidata/template_test.go:35]: Does the yaml library support unmarshalling this number literal to *string?

LGTM then, feel free to merge it

Comment (jandubois) [reply pkg/cidata/template_test.go:35]: Somewhat ironically the `UID` field in the cidata template is an `int`, and then gets turned into a string again in `user-data` with `uid: "{{.UID}}"`. 😄 

Anyways, I've change `uid` to an `int` in `lima.yaml` because I don't like changing the schema again later.

So please approve again once everything is green again.

Comment (jandubois) [reply pkg/cidata/template_test.go:35]: Ok, changed it one more time to `uint32` to avoid CodeQL errors...

Comment (jandubois) [reply pkg/cidata/template_test.go:35]: > LGTM then, feel free to merge it

You have requested changes; please formally approve. I don't want to merge without the explicit approval because it would hurt our OSSF score...

**Review by AkihiroSuda — APPROVED** [review]
Thanks

File: pkg/limayaml/limayaml.go
Comment (afbjorklund) [reply pkg/limayaml/limayaml.go]: You didn't like [General Electric Comprehensive Operating Supervisor](https://en.wikipedia.org/wiki/GECOS) ? :-)

Could have gone with `Info` as well, but since nobody looks at it anyway I think `Comment` is fine...

```C
           struct passwd {
               char   *pw_name;       /* username */
               char   *pw_passwd;     /* user password */
               uid_t   pw_uid;        /* user ID */
               gid_t   pw_gid;        /* group ID */
               char   *pw_gecos;      /* user information */
               char   *pw_dir;        /* home directory */
               char   *pw_shell;      /* shell program */
           };
```

Apparently I had some bugs with gecos in "tinycloudinit", and currently it uses `Linux User,,,`

https://github.com/mirror/busybox/blob/1_36_1/loginutils/adduser.c#L199

**Comment by afbjorklund** [comment]
> I'm wondering if `{{.Comment}}` should be supported in host/guest templates. It would be easy to add, but I can't think of an actual use-case.

I don't actually know the use case (even `finger` has been removed, for "pinky") - it was just that all the cloud-config examples had gecos...

**Pull request merged**
